README
SYNOPSIS
      use CGI::Application::Plugin::Stream 'stream';

      sub runmode {
        # ...

        # Set up any headers you want to set explicitly
        # using header_props() or header_add() as usual

        #...

        return $self->stream($file);
      }

DESCRIPTION
    This plugin provides a way to stream a file back to the user.

    This is useful if you are creating a PDF or Spreadsheet document
    dynamically to deliver to the user.

    The file is read and printed in small chunks to keep memory consumption
    down.

    The "stream()" method should be run as the last call in a a run mode,
    and affects the HTTP response headers. If you pass along a filehandle,
    we'll make sure to close it for you.

METHODS
  stream()
      $self->stream($fh);
      $self->stream( '/path/to/file',2048);

    This method can take two parameters, the first the path to the file or a
    filehandle and the second, an optional number of bytes to determine the
    chunk size of the stream. It defaults to 1024.

    It will either stream a file to the user or die if it fails, perhaps
    because it couldn't find the file you referenced.

    We highly recommend you provide a file name if passing along a
    filehandle, as we won't be able to deduce the file name, and will use
    'FILE' by default. Example:

     $self->header_add( -attachment => 'my_file.txt' );

    With both a file handle or file name, we will try to determine the
    correct content type by using File::MMagic. A default of
    'application/octet-stream' will be used if File::MMagic can't figure it
    out.

    The size will be calculated and added to the headers as well.

    Again, you can set these explicitly if you want as well:

     $self->header_add(
          -type                     =>      'text/plain',
          -Content_Length   =>      42, # bytes
     );

  stream_file()  ( deprecated )
    This works the same as stream(), but returns false to indicate a
    problem.

    This is included because this was the original API. It does not give you
    access to "$@" if there is a problem.

AUTHOR
    Jason Purdy, <Jason@Purdy.INFO>, with inspiration from Tobias Henoeckl
    and tremendous support from the cgiapp mailing list.

    Mark Stosberg also contributed to this module.

SEE ALSO
    CGI::Application, <http://www.cgi-app.org>, "CREATING A STANDARD HTTP
    HEADER" in CGI.pm,
    <http://www.mail-archive.com/cgiapp@lists.erlbaum.net/msg02660.html>,
    File::Basename, "$|" in perlvar

LICENSE
    Copyright (C) 2004-2005 Jason Purdy, <Jason@Purdy.INFO>

    This library is free software. You can modify and or distribute it under
    the same terms as Perl itself.



Changes
Revision history for Perl extension CGI::Application::Plugin::Stream.

2.11  Mon Dec 23 10:05:51 2013 -0500
  * Fixed header hyphen prefix that was causing testing failures
    - https://rt.cpan.org/Public/Bug/Display.html?id=81646
    - Thanks, Niko Tyni

2.10  Mon Mar 30 12:04:57 EDT 2009
  * Added support for CGI_APP_RETURN_ONLY env. variable
    - Thanks, Brad Bailey

2.06  Tue Dec 19 09:04:18 EST 2006
  * Testing library was missing some override functions
    - Thanks, Makio Tsukamoto
  * Minor whitespace issue
  * Added content regex to header check

2.05  Tue Sep 27 14:39:41 EDT 2005
    * Added README to MANIFEST (not sure how it got removed)

2.04  Thu Sep  8 11:52:34 EDT 2005
    * Added test coverage for the PODs, to boost Kwalitee

2.03  Sun Jul  3 22:13:00 2005
    [BUG FIX]
    - Patch - tolerates absence of File::MMagic (Thilo Phanz)

2.02  Thu Jun 23 17:01:00 2005
    [BUG FIX]
    - Added a binmode line to address a problem on Windows (William McKee)

2.01  Wed Mar 30 14:59:17 2005
    [BUG FIXES]
    - We now use the correct 'Content-Length' header instead of 'Size'. (Mark Stosberg)
    - A case in which the first few characters of the file could have been
      missing after the stream was fixed. (Mark Stosberg) 

2.00  Mon Dec  6 14:26:17 2004
    The use of the second argument to stream_file() has changed and is not
    backward compatible.

    [ENHANCEMENTS]
    - header_props() and header_add() can now be used to manage headers
    - The right content type will now attempt to be auto-detected if none is
      provided.
    - The chunk size of the stream is now configurable. 

    [INTERNALS]
    - A test suite was added.
    - started Exporter import technique is used now.
    - Documentation was updated
    - Test::More, File::MMagic and FileHandle are now required. 
    - Migrate distribution management to Module::Build

1.00  Thu Dec  3 10:42:04 2004
	- production version

0.01  Thu Dec  2 16:07:45 2004
	- original version; created by h2xs 1.21 with options
		-AX -n CGI::Application::Plugin::Stream



lib/CGI/Application/Plugin/Stream.pm
package CGI::Application::Plugin::Stream;

use 5.006;
use strict;
use warnings;

use CGI::Application 3.21;
use File::Basename;
require Exporter;
use vars (qw/@ISA @EXPORT_OK/);
@ISA = qw(Exporter);

@EXPORT_OK = qw(stream_file);

our $VERSION = '2.11';

sub stream_file {
    my ( $self, $file_or_fh, $bytes ) = @_;
    $bytes ||= 1024;
    my ($fh, $basename);
    my  $size = (stat( $file_or_fh ))[7];

    # If we have a file path
    if ( ref( \$file_or_fh ) eq 'SCALAR' ) {
        # They passed along a scalar, pointing to the path of the file
        # So we need to open the file
        open($fh,"<$file_or_fh"  ) || return 0;
        # Now let's go binmode (Thanks, William!)
        binmode $fh;
        $basename = basename( $file_or_fh );
    }
    # We have a file handle.
    else {
        $fh = $file_or_fh;
        $basename = 'FILE';
    }

    # Use FileHandle to make File::MMagic happy;
    # bless the filehandle into the FileHandle package to make File::MMagic happy
    require FileHandle;
    bless $fh,  "FileHandle";

    # Check what headers the user has already set and
    # don't override them.
    my %existing_headers = $self->header_props();

    # Check for a existing type header set with or without a hyphen
    unless ( $existing_headers{'-type'} ||  $existing_headers{'type'} ) {
        my $mime_type;

        eval {
            require File::MMagic;
            my $magic = File::MMagic->new();
            $mime_type = $magic->checktype_filehandle($fh);
        };
        warn "Failed to load File::MMagic module to determine mime type: $@" if $@;

        # Set Default
        $mime_type ||= 'application/octet-stream';

        $self->header_add('-type' => $mime_type);
    }


    unless ( $existing_headers{'Content_Length'}
        ||   $existing_headers{'-Content_Length'}
        ) {
        $self->header_add('-Content_Length' => $size);
    }

    unless ( $existing_headers{'-attachment'}
        ||   $existing_headers{'attachment'}
        ||   grep( /-?content-disposition/i, keys %existing_headers )
        ) {
        $self->header_add('-attachment' => $basename);
    }

    unless ( $ENV{'CGI_APP_RETURN_ONLY'} ) {
        $self->header_type( 'none' );
        print $self->query->header( $self->header_props() );
    }

    # This reads in the file in $byte size chunks
    # File::MMagic may have read some of the file, so seek back to the beginning
    my $output = "";
    seek($fh,0,0);
    while ( read( $fh, my $buffer, $bytes ) ) {
        if ( $ENV{'CGI_APP_RETURN_ONLY'} ) {
            $output .= $buffer;
        } else {
            print $buffer;
        }
    }

    print '' unless $ENV{'CGI_APP_RETURN_ONLY'}; # print a null string at the end
    close ( $fh );
    return $ENV{'CGI_APP_RETURN_ONLY'} ? \$output : 1;
}

1;
__END__
=head1 NAME

CGI::Application::Plugin::Stream - CGI::Application Plugin for streaming files

=head1 SYNOPSIS

  use CGI::Application::Plugin::Stream (qw/stream_file/);

  sub runmode {
    # ...

    # Set up any headers you want to set explicitly
    # using header_props() or header_add() as usual

    #...

    if ( $self->stream_file( $file ) ) {
      return;
    } else {
      return $self->error_mode();
    }
  }

=head1 DESCRIPTION

This plugin provides a way to stream a file back to the user.

This is useful if you are creating a PDF or Spreadsheet document dynamically to
deliver to the user.

The file is read and printed in small chunks to keep memory consumption down.

This plugin is a consumer, as in your runmode shouldn't try to do any output or
anything afterwards.  This plugin affects the HTTP response headers, so
anything you do afterwards will probably not work.  If you pass along a
filehandle, we'll make sure to close it for you.

It's recommended that you increment $| (or set it to 1), which will
autoflush the buffer as your application is streaming out the file.

=head1 METHODS

=head2 stream_file()

  $self->stream_file($fh);
  $self->stream_file( '/path/to/file',2048);

This method can take two parameters, the first the path to the file
or a filehandle and the second, an optional number of bytes to determine
the chunk size of the stream. It defaults to 1024.

It will either stream a file to the user or return false if it fails, perhaps
because it couldn't find the file you referenced.

We highly recommend you provide a file name if passing along a filehandle, as we
won't be able to deduce the file name, and will use 'FILE' by default. Example:

 $self->header_add( -attachment => 'my_file.txt' );

With both a file handle or file name, we will try to determine the correct
content type by using File::MMagic. A default of 'application/octet-stream'
will be used if File::MMagic can't figure it out.

The size will be calculated and added to the headers as well.

Again, you can set these explicitly if you want as well:

 $self->header_add(
      -type		        =>	'text/plain',
      -Content_Length	=>	42, # bytes
 );

=head1 AUTHOR

Jason Purdy, E<lt>Jason@Purdy.INFOE<gt>,
with inspiration from Tobias Henoeckl
and tremendous support from the cgiapp mailing list.

Mark Stosberg also contributed to this module.

=head1 SEE ALSO

L<CGI::Application>,
L<http://www.cgi-app.org>,
L<CGI.pm/"CREATING A STANDARD HTTP HEADER">,
L<http://www.mail-archive.com/cgiapp@lists.erlbaum.net/msg02660.html>,
L<File::Basename>,
L<perlvar/$E<verbar>>

=head1 LICENSE

Copyright (C) 2004-2005 Jason Purdy, E<lt>Jason@Purdy.INFOE<gt>

This library is free software. You can modify and or distribute it under the same terms as Perl itself.

=cut
